crates/governance/src/finalize_block.rs

```diff
@@ -0,0 +1,656 @@
+//! Governance logic applied on an end of a block.
+
+use std::borrow::Cow;
+use std::collections::BTreeSet;
+use std::fmt::Display;
+use std::str::FromStr;
+
+use borsh::BorshDeserialize;
+use namada_core::address::Address;
+use namada_core::chain::{ChainId, Epoch};
+use namada_core::collections::HashMap;
+use namada_core::encode;
+use namada_core::ibc::PGFIbcTarget;
+use namada_events::extend::{ComposeEvent, Height};
+use namada_events::{EmitEvents, EventLevel};
+use namada_state::{Key, Result, State, StateRead, StorageRead, StorageWrite};
+use namada_systems::{proof_of_stake, trans_token as token};
+use namada_tx::data::TxType;
+use namada_tx::{Code, Data, Tx};
+
+use crate::event::GovernanceEvent;
+use crate::pgf::storage::keys as pgf_keys;
+use crate::pgf::storage::steward::StewardDetail;
+use crate::pgf::{storage as pgf_storage, ADDRESS as PGF_ADDRESS};
+use crate::storage::proposal::{
+    AddRemove, PGFAction, PGFTarget, ProposalType, StoragePgfFunding,
+};
+use crate::storage::{keys, load_proposals};
+use crate::utils::{
+    compute_proposal_result, ProposalVotes, TallyResult, TallyType, VotePower,
+};
+use crate::{storage, ProposalVote, ADDRESS as GOV_ADDRESS};
+
+/// Apply governance updates for a block. On a new epoch, this will look for
+/// proposals to tally completed proposals and execute accepted proposals.
+#[allow(clippy::too_many_arguments)]
+pub fn finalize_block<S, Token, PoS, FnTx, VpCache, TxCache, FnIbcTransfer>(
+    state: &mut S,
+    vp_wasm_cache: &mut VpCache,
+    tx_wasm_cache: &mut TxCache,
+    events: &mut impl EmitEvents,
+    current_epoch: Epoch,
+    is_new_epoch: bool,
+    dispatch_tx: FnTx,
+    transfer_over_ibc: FnIbcTransfer,
+) -> Result<Option<ProposalsResult>>
+where
+    S: StateRead + State,
+    Token: token::Read<S> + token::Write<S> + token::Events<S>,
+    PoS: proof_of_stake::Read<S>,
+    FnTx: FnMut(&Tx, &mut S, &mut VpCache, &mut TxCache) -> Result<bool>,
+    FnIbcTransfer: Fn(&mut S, &Address, &Address, &PGFIbcTarget) -> Result<()>,
+{
+    if is_new_epoch {
+        let result = load_and_execute_governance_proposals::<
+            S,
+            Token,
+            PoS,
+            FnTx,
+            VpCache,
+            TxCache,
+            FnIbcTransfer,
+        >(
+            state,
+            vp_wasm_cache,
+            tx_wasm_cache,
+            events,
+            current_epoch,
+            dispatch_tx,
+            transfer_over_ibc,
+        )?;
+        if result.passed.is_empty() && result.rejected.is_empty() {
+            return Ok(None);
+        }
+        return Ok(Some(result));
+    }
+    Ok(None)
+}
+
+/// Count of passed and rejected proposals.
+#[derive(Default)]
+pub struct ProposalsResult {
+    passed: Vec<u64>,
+    rejected: Vec<u64>,
+}
```

Comment:
maybe we can just return `Result<(), ..>` ? I dont think we need to return to the called which proposal passed and which didn't?

Reply:
sounds good

Reply:
done in [ae95058](https://github.com/anoma/namada/pull/3718/commits/ae950581613927fff73765ab7ed19b221d78286e)